### PR TITLE
Sema: Warn when Swift classes attempt to implement ObjC +initialize.

### PR DESCRIPTION
**Pull request by jckarter**
Swift's language model doesn't guarantee that type metadata will ever really be used, which makes overriding initialize() error-prone and not really any better than manually invoking an initialization function. Warn about this for Swift 3 compatibility and reject attempts to override +initialize in Swift 4.

### CODE REVIEW

**Comment by jckarter** [comment]
@swift-ci Please smoke test

**Comment by jckarter** [comment]
@jrose-apple Does this look reasonable for 3.1?

**Review by DougGregor — APPROVED** [review]
LGTM!

File: include/swift/AST/DiagnosticsSema.def [hunk 1]
```diff
@@ -3137,6 +3137,11 @@ ERROR(objc_class_method_not_permitted,none,
       OBJC_DIAG_SELECT " defines Objective-C class method %2, which is "
       "not permitted by Swift", (unsigned, DeclName, ObjCSelector))
 
+WARNING(objc_class_method_not_permitted_swift3_compat_warning,none,
+      OBJC_DIAG_SELECT " defines Objective-C class method %2, which is "
+      "not guaranteed to be invoked by Swift and will be disallowed in future "
```
Comment (slavapestov) [inline include/swift/AST/DiagnosticsSema.def]: Nitpick: indentation

**Pull request merged**